docs/rfc/rfc-023-semi-permanent-testnet.md

```diff
@@ -0,0 +1,206 @@
+
+In summary, if the tendermint authors build, maintain, and continuously deliver an application
+intended to serve as a long-lived testnet, they will be able to state with confidence:
+- We operate the software in a production-like environment and we have observed it to be
+  stable and performant to our requirements
+- We have discovered issues in production before any external parties have consumed our software,
+  and we have addressed said issues
+- We have successfully used the observability tooling built into our software
+  (perhaps in conjunction with other off-the-shelf tooling)
+  to diagnose and debug issues in production
+
+## Discussion
+
+The Discussion Section proposes a variety of aspects of maintaining a testnet for Tendermint.
+
+### Number of testnets
+
+There should probably be one testnet per maintained branch of Tendermint,
+i.e. one for the `main` branch
+and one per `v0.N.x` branch that the authors maintain.
+
+There may also exist testnets for long-lived feature branches.
+
+We may eventually discover that there is good reason to run more than one testnet for a branch,
+perhaps due to a significant configuration variation.
+
+### Testnet lifecycle
+
+The document has used the terms "long-lived" and "semi-permanent" somewhat interchangeably.
+The intent of the testnet being discussed in this RFC is to exist indefinitely;
+but there is a practical understanding that there will be testnet instances
+which will be retired due to being associated with a no-longer-maintained release of Tendermint.
+Alternatively, we may discover a good reason to destroy and recreate the `main` testnet
+(such as if we build a new application for running the testnet).
+
+### Testnet MVP
+
+The minimum viable testnet meets this set of features:
+
+- The testnet self-updates following a new commit pushed to Tendermint's `main` branch on GitHub
+  (there are some omitted steps here, such as CI building appropriate binaries and
+  somehow notifying the testnet that a new build is available)
+- The testnet runs the Tendermint KV store for MVP
+- The testnet operators are notified if any node's process exits for any reason other than
+  a restart for a new binary
+- The testnet operators are notified if any node stops updating blocks,
+  and by extension if a chain halt occurs
+- The testnet has a minimum of 1 full node and 3 validators
+- The testnet has a reasonably low, constant throughput of transactions -- say 30 tx/min --
+  and the testnet operators are notified if that throughput drops below 75% of target
+  sustained over 5 minutes
+- The testnet only needs to run in a single datacenter/cloud-region for MVP,
+  i.e. running in multiple datacenters is out of scope for MVP
+- The testnet is running directly on VMs or compute instances;
+  while Kubernetes or other orchestration frameworks may offer many significant advantages,
+  the Tendermint engineers should not be required to learn those tools in order to
+  perform basic debugging
```

Comment:
I would contend that using kube, in general, should be a non-goal of any such project. Our users almost exclusively do _not_ use it from what I've observed. They run the binary on VMs and we should attempt to mimic what they do as much as possible.

Reply:
I've heard that the Cosmos ecosystem is a decent mix where the "smaller" validators tend to run directly on VMs, and the "larger" ones tend to run in containerized environments. But I don't have any hard data to back that up.

If that is accurate, I think there is some value **in the long term** in running Tendermint nodes in containers.

In either case, I am comfortable beginning without any plans for containerized nodes.